# Part 1. Equation of a Slime

How many late days are you using for this assignment?
0

In [44]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score, KFold
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier
import numpy as np

## 1. Loading the dataset

In [45]:
# Using pandas load the dataset
alien_df = pd.read_csv('science_data_large.csv')
# Output the first 15 rows of the data
print(alien_df[:15])
# Display a summary of the table information (number of datapoints, etc.)
print(alien_df.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [46]:
# Take the pandas dataset and split it into our features (X) and label (y)
alien_X = alien_df[['Mols KCL', 'Temperature °C']]
alien_y = alien_df['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
alien_X_train, alien_X_test, alien_y_train, alien_y_test = train_test_split(
    alien_X, alien_y, test_size=0.1, random_state=42)

## 3. Perform a Linear Regression

In [47]:
# Use sklearn to train a model on the training set
alien_reg = LinearRegression().fit(alien_X_train, alien_y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
print(alien_reg.predict([[800, 100]]))
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
print(alien_reg.score(alien_X_test, alien_y_test))
# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX
print(alien_reg.coef_)
print(alien_reg.intercept_)

[503379.21494242]
0.8552472077276096
[1032.69506649  866.14641337]
-409391.4795834087


c:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Write the linear equation of a slime: (example equation: $E = mc^2$)

Let 'Mols KCL' = M, 'Temperature °C' = T, 'Size nm^3' = S

$S = 1032.7M + 866.1T -409391.5$

Report on score and explain meaning:

Score uses the ratio of sum of squared residuals and the sum of squares (which measures variability in data) to give the model a score of 1 or below, where 1 is a perfect score. Our model got a score of 0.86, which is fairly close to a score of 1 and indicates that the model performs reasonably well.

## 4. Use Cross Validation

In [48]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
kfoldcv = KFold(n_splits=5, random_state=42, shuffle=True)

alien_scores = cross_val_score(alien_reg, alien_X, alien_y, cv=kfoldcv)
# For grading consistency use n_splits=5 and random_state=42
# Report on their finding and their significance
print(alien_scores)
print(alien_scores.mean())
print(alien_scores.std())

[0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
0.8597294202684646
0.01838773713930647


Write findings here:

Linear regression consistently performs quite well on the data, acheiving an average score of 0.86 and standard deviation of 0.02. This is consistent with the results of the original model.

## 5. Using Polynomial Regression

In [49]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
alien_poly = poly.fit_transform(alien_X)

alien_poly_reg = LinearRegression().fit(alien_poly, alien_y)
# Perform k-fold cross validation (as above)
alien_poly_scores = cross_val_score(alien_poly_reg, alien_poly, alien_y, cv=kfoldcv)

# print(alien_poly_reg.predict([alien_poly[1]]))

# Report on the metrics and output the resultant equation as you did in Part 3.

print(alien_poly_reg.coef_)
print(alien_poly_reg.intercept_)

print(alien_poly_scores)
print(alien_poly_scores.mean())
print(alien_poly_scores.std())

[ 0.00000000e+00 -1.23112976e-07  1.20000000e+01  2.85714287e-02
  2.00000000e+00 -1.05628999e-11]
1.6571662854403257e-05
[1. 1. 1. 1. 1.]
1.0
0.0


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

Let 'Mols KCL' = M, 'Temperature °C' = T, 'Size nm^3' = S

$S = -0.00000012M + 12T + 0.0286M^2 + 2MT - 0.000000000011B^2$

Report on the score and interpret:

The scores were consistent ones across the board, indicating that the model performed perfectly on the data accross all folds. This, to me, indicates that the size of the slime is a direct, nonlinear function of temperature and potassium chloride exposure.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [50]:
# Using pandas load the dataset
ckd_df = pd.read_csv('ckd_feature_subset.csv')
# Output the first 15 rows of the data
print(ckd_df[:15])
# Display a summary of the table information (number of datapoints, etc.)
print(ckd_df.info())

         age        bp      wbcc  appet_poor  appet_good      rbcc  Target_ckd
0   0.688312  0.333333  0.000000           1           0  0.000000           1
1   0.545455  0.333333  0.128319           1           0  0.305085           1
2   0.714286  0.500000  0.238938           1           0  0.186441           1
3   0.688312  0.333333  0.283186           0           1  0.338983           1
4   0.441558  0.333333  0.221239           1           0  0.220339           1
5   0.740260  0.833333  0.265487           0           1  0.355932           1
6   0.870130  0.333333  0.668142           0           1  0.237288           1
7   0.870130  0.500000  0.150442           0           1  0.372881           1
8   0.194805  0.666667  0.380531           0           1  0.305085           1
9   0.753247  0.833333  0.163717           0           1  0.220339           1
10  0.701299  0.166667  0.504425           1           0  0.152542           1
11  0.610390  0.666667  0.367257           1        

In [51]:
# Take the pandas dataset and split it into our features (X) and label (y)
ckd_X = ckd_df[['age', 'bp', 'wbcc', 'appet_poor', 'appet_good', 'rbcc']]
ckd_y = ckd_df['Target_ckd']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
ckd_X_train, ckd_X_test, ckd_y_train, ckd_y_test = train_test_split(
    ckd_X, ckd_y, test_size=0.1, random_state=42)

In [52]:
# Use sklearn to train a model on the training set
ckd_reg = LogisticRegression().fit(ckd_X_train, ckd_y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
print(ckd_reg.score(ckd_X_test, ckd_y_test))
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
kfoldcv = KFold(n_splits=5, random_state=42, shuffle=True)

ckd_scores = cross_val_score(ckd_reg, ckd_X, ckd_y, cv=kfoldcv)
# For grading consistency use n_splits=5 and random_state=42
# Report on their finding and their significance
print(ckd_scores)
print(ckd_scores.mean())
print(ckd_scores.std())

0.6875
[0.80645161 0.93548387 0.77419355 0.83333333 0.93333333]
0.8565591397849461
0.0662686697755876


In [53]:
# 1. Train and test a Support Vector Machine (SVM) model
ckd_svm = SVC(random_state=42).fit(ckd_X_train, ckd_y_train)
print("SVM Test Score:", ckd_svm.score(ckd_X_test, ckd_y_test))

kfoldcv = KFold(n_splits=5, random_state=42, shuffle=True)
svm_scores = cross_val_score(ckd_svm, ckd_X, ckd_y, cv=kfoldcv)
print("SVM Cross-Validation Scores:", svm_scores)
print("Mean Score:", svm_scores.mean())
print("Standard Deviation:", svm_scores.std())

SVM Test Score: 0.875
SVM Cross-Validation Scores: [0.90322581 1.         0.87096774 0.9        0.96666667]
Mean Score: 0.9281720430107526
Standard Deviation: 0.04760126519964333


In [54]:
# 2. Train and test a K-Nearest Neighbors (KNN) model
ckd_knn = KNeighborsClassifier().fit(ckd_X_train, ckd_y_train)
print("KNN Test Score:", ckd_knn.score(ckd_X_test, ckd_y_test))

kfoldcv = KFold(n_splits=5, random_state=42, shuffle=True)
knn_scores = cross_val_score(ckd_knn, ckd_X, ckd_y, cv=kfoldcv)
print("KNN Cross-Validation Scores:", knn_scores)
print("Mean Score:", knn_scores.mean())
print("Standard Deviation:", knn_scores.std())

KNN Test Score: 0.9375
KNN Cross-Validation Scores: [0.93548387 1.         0.87096774 0.86666667 0.96666667]
Mean Score: 0.9279569892473118
Standard Deviation: 0.05243961560473918


In [55]:
# 3. Train and test a Neural Network model
ckd_nn = MLPClassifier(random_state=42).fit(ckd_X_train, ckd_y_train)
print("Neural Network Test Score:", ckd_nn.score(ckd_X_test, ckd_y_test))

kfoldcv = KFold(n_splits=5, random_state=42, shuffle=True)
nn_scores = cross_val_score(ckd_nn, ckd_X, ckd_y, cv=kfoldcv)
print("Neural Network Cross-Validation Scores:", nn_scores)
print("Mean Score:", nn_scores.mean())
print("Standard Deviation:", nn_scores.std())

c:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Neural Network Test Score: 0.75
Neural Network Cross-Validation Scores: [0.83870968 0.93548387 0.77419355 0.86666667 0.96666667]
Mean Score: 0.8763440860215054
Standard Deviation: 0.06870286199728586


c:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Classification Experiments

| Model        | Mean Score | Std      |
|--------------|------------|----------|
| LogReg       | 0.86       | 0.07     |
| SVM          | 0.93       | 0.05     |
| KNN          | 0.93       | 0.05     |
| neuralNet    | 0.88       | 0.07     |

Logistic Regression:

The scores for the logistic regression model are a bit inconsistent. The mean score is a .86, which is good, but there is a standard deviation of 0.07

Support Vector Machines:

The scores using support vector machines were much more promising, with a mean score of 0.93 and a standard deviation of five. This means that the model performed more accurately and more consistently than regression,

k-Nearest Neighbors:

The KNN algorithm produced almost identical results to support vector machines.

Neural Network:

The neural network provided similar results to logistic regression. I did, however, receive a warning that the optimization has not converged after the maximum number of iterations was reached. This implies that the neural network may still be promising given a greater number of trainig epochs.

In [69]:
ckd_nn = MLPClassifier(max_iter=2000, hidden_layer_sizes=(100, 50, 25,), learning_rate='adaptive', random_state=42).fit(ckd_X_train, ckd_y_train)
print("Neural Network Test Score:", ckd_nn.score(ckd_X_test, ckd_y_test))

kfoldcv = KFold(n_splits=5, random_state=42, shuffle=True)
nn_scores = cross_val_score(ckd_nn, ckd_X, ckd_y, cv=kfoldcv)
print("Neural Network Cross-Validation Scores:", nn_scores)
print("Mean Score:", nn_scores.mean())
print("Standard Deviation:", nn_scores.std())

Neural Network Test Score: 0.9375
Neural Network Cross-Validation Scores: [0.96774194 1.         0.90322581 1.         0.96666667]
Mean Score: 0.9675268817204301
Standard Deviation: 0.03533955664799394


In [70]:
ckd_nn = MLPClassifier(max_iter=2000, random_state=42).fit(ckd_X_train, ckd_y_train)
print('More Training Neural Net:')
print("Neural Network Test Score:", ckd_nn.score(ckd_X_test, ckd_y_test))

kfoldcv = KFold(n_splits=5, random_state=42, shuffle=True)
nn_scores = cross_val_score(ckd_nn, ckd_X, ckd_y, cv=kfoldcv)
print("Neural Network Cross-Validation Scores:", nn_scores)
print("Mean Score:", nn_scores.mean())
print("Standard Deviation:", nn_scores.std())

More Training Neural Net:
Neural Network Test Score: 0.875
Neural Network Cross-Validation Scores: [0.90322581 0.96774194 0.87096774 0.96666667 0.96666667]
Mean Score: 0.9350537634408603
Standard Deviation: 0.040465552982107675


In [71]:
ckd_nn = MLPClassifier(learning_rate='adaptive', random_state=42).fit(ckd_X_train, ckd_y_train)
print('Adaptive Learning Neural Net:')
print("Neural Network Test Score:", ckd_nn.score(ckd_X_test, ckd_y_test))

kfoldcv = KFold(n_splits=5, random_state=42, shuffle=True)
nn_scores = cross_val_score(ckd_nn, ckd_X, ckd_y, cv=kfoldcv)
print("Neural Network Cross-Validation Scores:", nn_scores)
print("Mean Score:", nn_scores.mean())
print("Standard Deviation:", nn_scores.std())

c:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Adaptive Learning Neural Net:
Neural Network Test Score: 0.75
Neural Network Cross-Validation Scores: [0.83870968 0.93548387 0.77419355 0.86666667 0.96666667]
Mean Score: 0.8763440860215054
Standard Deviation: 0.06870286199728586


c:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


In [ ]:
ckd_nn = MLPClassifier(hidden_layer_sizes=(100, 50, 25,), random_state=42).fit(ckd_X_train, ckd_y_train)
print('Larger Neural Net:')
print("Neural Network Test Score:", ckd_nn.score(ckd_X_test, ckd_y_test))

kfoldcv = KFold(n_splits=5, random_state=42, shuffle=True)
nn_scores = cross_val_score(ckd_nn, ckd_X, ckd_y, cv=kfoldcv)
print("Neural Network Cross-Validation Scores:", nn_scores)
print("Mean Score:", nn_scores.mean())
print("Standard Deviation:", nn_scores.std())

c:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Larger Neural Net:
Neural Network Test Score: 0.9375


c:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


Neural Network Cross-Validation Scores: [0.93548387 1.         0.90322581 1.         0.96666667]
Mean Score: 0.9610752688172044
Standard Deviation: 0.03758460606443611


c:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
c:\Users\<redacted>\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\neural_network\_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Neural Network Experiments

| Model        | Mean Score | Std      |
|--------------|------------|----------|
| More Epochs  | 0.93       | 0.04     |
| ALR          | 0.88       | 0.07     |
| Larger       | 0.96       | 0.04     |
| Normal       | 0.88       | 0.07     |

More Epochs:

The neural net trained for more epochs, which had max_iter set to 2000, performed better than the initial model. It appears I was correct that more training would benefit the model. No other parameters were changed.

ALR:

Setting learning_rate to 'adaptive' provided no clear improvement or deterioration of the model's performance.

Larger:

The larger neural network, with hidden_layer_sizes set to (100, 50, 25,), performed the best with a mean score of 0.96. Larger models typically display diminishing returns, and it is important the think about the balance between model size and training/prediction time/cost.
